Incorporate `gempyor.distributions` into `gempyor.parameters`

emprzy opened the pull request:
### Describe your changes.

This pull request updates the `gempyor.parameters` submodule to use the distributions recently defined in `gempyor.distributions`. It also updates the tests in `test_parameters_class.py` accordingly.


### Does this pull request make any user interface changes? If so please describe.

There are no user interface changes.


### What does your pull request address? Tag relevant issues.

This pull request addresses GH #582

emprzy committed:
Incorporate `gempyor.distributions` into `gempyor.parameters`

emprzy committed:
Update `Parameters` class testing

Update the `test_parameters_class.py` file to be up to date with new `gempyor.distributions` incorporation.

emprzy commented on flepimop/gempyor_pkg/tests/parameters/test_parameters_class.py:
```diff
@@ -343,37 +362,45 @@ def test_parameters_instance_attributes(
+                dist_obj = params.pdata[param_name]["dist"]
+                value_conf = param_conf.get("value")
+
+                # TODO, does this just test pydantic?
```
I have a question here, I am having a hard time parsing if this is test logic that shouldn't exist (is it just testing Pydantic stuff)?

pearsonca replied:
yeah, this is just testing that dict got parsed

emprzy commented:
Okay, I actually think I did this incorporation too obtrusively (based on how many tests need updating). Going back to re-work.

emprzy committed:
Fix bugs created in last push

emprzy committed:
Remove unnecessary assertions in `test_parameters_class.py`

We don't need to test vendor libraries! I need to write that 1,000 times on a piece of paper or something haha

pearsonca commented on flepimop/gempyor_pkg/src/gempyor/parameters.py:
```diff
@@ -97,9 +99,12 @@ def __init__(
             if self.pconfig[pn]["value"].exists():
-                self.pdata[pn]["dist"] = self.pconfig[pn]["value"].as_random_distribution()
+                dist_dict = self.pconfig[pn]["value"].get()
+                if isinstance(dist_dict, float | int | str):
+                    dist_dict = {"distribution": "fixed", "value": dist_dict}
```
this is relying on internal details of how distributions work - let's relocate this translation into distributions

TimothyWillard replied:
To add to this, I agree, but `pydantic.TypeAdapter` does not provide a great hook to preprocess a given value. So the solution here will be to add a small helper to `gempyor.distributions` called something like `distribution_from_confuse_config` (or similar) that holds this logic. I.e.

```python
def distribution_from_confuse_config(param_cfg: confuse.ConfigView) -> DistributionABC:
  conf = param_cfg["value"].get()
  if isinstance(conf, float | int | str):
    conf = {"distribution": "fixed", "value": conf}
  return DISTRIBUTION_ADAPTER.validate_python(conf)
```

pearsonca replied:
yes, but: i recall ConfigView `get()` has some slightly strange behavior - best double check what it's actual return is.

emprzy replied:
@TimothyWillard @pearsonca See most recent commit to ensure I executed this according to your vision.

pearsonca commented on flepimop/gempyor_pkg/src/gempyor/parameters.py:
```diff
@@ -224,7 +232,9 @@ class via `subpop_names`.
-                param_arr[idx] = np.full((n_days, nsubpops), self.pdata[pn]["dist"]())
+                param_arr[idx] = np.full(
+                    (n_days, nsubpops), self.pdata[pn]["dist"].sample().item()
+                )
```
this is a faithful translation, but - are we meaning to replicate the value here, rather than draw for the structure? let's double check

TimothyWillard replied:
This is correct. The `param_arr` has dimension `(# parameters, # days, # subpopulations)`. If we were to draw for the whole structure this would be like having a different and independently drawn parameter value for each day/subpopulation. Drawing one value and replicating corresponds to one parameter draw. I think ultimately we want to change this structure to provide users some more flexibility (i.e. #469, #438), but that is outside the scope of this PR.

pearsonca commented on flepimop/gempyor_pkg/src/gempyor/parameters.py:
```diff
@@ -272,8 +282,9 @@ def parameters_load(
                 )
-                pval = self.pdata[pn]["dist"]()
-                param_arr[idx] = np.full((n_days, nsubpops), pval)
+                param_arr[idx] = np.full(
```
ibid

pearsonca commented on flepimop/gempyor_pkg/src/gempyor/parameters.py:
```diff
@@ -182,7 +187,10 @@ def reinitialize_distributions(self) -> None:
             if "dist" in self.pdata[pn]:
-                self.pdata[pn]["dist"] = self.pconfig[pn]["value"].as_random_distribution()
+                dist_dict = self.pconfig[pn]["value"].get()
+                if isinstance(dist_dict, float | int | str):
```
ibid

pearsonca commented on flepimop/gempyor_pkg/tests/parameters/test_parameters_class.py:
```diff
@@ -467,20 +460,48 @@ def test_parameters_quick_draw(
+                    drawn_values = p_draw[i, :, :]
+
+                    # For a fixed distribution, all values must be equal to the specified value
+                    if isinstance(dist_obj, FixedDistribution):
```
we already test that FixedDistribution et al work, right? what distinctly are we checking here?

emprzy replied:
I believe this is resolved by Tim's suggestion, [113a8bb](https://github.com/HopkinsIDD/flepiMoP/pull/584/commits/113a8bbe9986626e29c4e0e4980643051a331734)

pearsonca commented on flepimop/gempyor_pkg/tests/parameters/test_parameters_class.py:
```diff
@@ -599,28 +620,56 @@ def test_parameters_load(
+                        assert np.allclose(drawn_values, dist_obj.value)
+
+                    # For bounded distributions, check that all values are within the domain
+                    elif isinstance(dist_obj, UniformDistribution):
```
ibid

emprzy replied:
I believe this is resolved by Tim's suggestion, [113a8bb](https://github.com/HopkinsIDD/flepiMoP/pull/584/commits/113a8bbe9986626e29c4e0e4980643051a331734)

TimothyWillard replied:
Are these assertions for `drawn_values` still needed? I think it would be sufficient to do `isinstance(dist_obj, DistributionABC)`.

pearsonca reviewed (changes requested):
Feels like this needs more. Also, definitely need to test the eval expression mechanism - probably by extending tests elsewhere for either generic pydantic capabilities or at least distributions.

TimothyWillard commented on flepimop/gempyor_pkg/src/gempyor/_pydantic_ext.py:
```diff
@@ -180,3 +182,31 @@ def _read_and_validate_dataframe(
     return pd.DataFrame.from_records(data)
+
+
+def _evaled_expression(val: float | str | Any) -> float | Any:
```
Requires unit tests, an `Examples:` section would be helpful too. Also it would be good to add the `@overload`s for this annotation.

TimothyWillard commented on flepimop/gempyor_pkg/src/gempyor/_pydantic_ext.py:
```diff
@@ -180,3 +182,31 @@ def _read_and_validate_dataframe(
+    return val
+
+
+# TODO: add EvaluatedInt = Annotated[int, BeforeValidator(_evaled_expression_int)]
```
This needs to be implemented. Might also want to think about how to unify hypothetical `_evaled_expression_int` and existing `_evaled_expression` into one function.

TimothyWillard commented on flepimop/gempyor_pkg/tests/parameters/test_parameters_class.py:
```diff
@@ -467,20 +460,48 @@ def test_parameters_quick_draw(
+                    # For distributions without a simple domain to check (Normal)
+                    else:
+                        assert isinstance(dist_obj, NormalDistribution)
+                        assert not np.isnan(drawn_values).any()
```
```suggestion
```

I think these should be covered by the distributions unit tests.

TimothyWillard commented on flepimop/gempyor_pkg/tests/parameters/test_parameters_class.py:
```diff
@@ -599,28 +620,56 @@ def test_parameters_load(
                 i = params.pnames.index(param_name)
+
                 if param_name in param_df["parameter"].values:
-                    # Check that the values in p_draw[i, :, :] match override
```
Minor: Are these comments not helpful?

emprzy replied:
I suppose they are, we just generally have pretty limited comments in our test functions so I was trying to stick to that aesthetic. I usually put a lot in when I write code and then delete them before I commit, but I can put these back if you want.

TimothyWillard commented on flepimop/gempyor_pkg/src/gempyor/_pydantic_ext.py:
```diff
@@ -180,3 +182,31 @@ def _read_and_validate_dataframe(
+
+    if isinstance(val, float):
+        return val
+    elif isinstance(val, str):
```
```suggestion
    if isinstance(val, str):
```

emprzy committed:
Update flepimop/gempyor_pkg/tests/parameters/test_parameters_class.py

Co-authored-by: Timothy Willard <9395586+TimothyWillard@users.noreply.github.com>

emprzy committed:
Update flepimop/gempyor_pkg/src/gempyor/_pydantic_ext.py

Co-authored-by: Timothy Willard <9395586+TimothyWillard@users.noreply.github.com>

emprzy committed:
Improvements surrounding new `_evaled_expression()` function

Add `@overload` ext for `_evaled_expression`, improve documentation and add unit testing

emprzy commented:
@TimothyWillard, something that occurred to me was that perhaps it would be nice to replace uses of the `gempyor.utils::as_evaled_expression()` function with the new Pydantic ext. `_evaled_expression()` function. Thoughts? Perhaps it is not ideal because it is not ~exactly the same implementation, but it's pretty close. Let me know what you think.

TimothyWillard commented:
> @TimothyWillard, something that occurred to me was that perhaps it would be nice to replace uses of the `gempyor.utils::as_evaled_expression()` function with the new Pydantic ext. `_evaled_expression()` function. Thoughts? Perhaps it is not ideal because it is not ~exactly~ the same implementation, but it's pretty close. Let me know what you think.

Yeah, I think this is a good idea, the implementation differences are minimal right? Glancing at the function bodies I do not see a difference. Here is where we'd have to replace our custom confuse addition with a regular function call, i.e. `_evaled_expression(self["value"].get())`:

```
> rg 'as_evaled_expression' src/gempyor/
src/gempyor/utils.py
354:def as_evaled_expression(self):
554:                self["value"].as_evaled_expression(),
555:                self["value"].as_evaled_expression(),
560:                self["low"].as_evaled_expression(),
561:                self["high"].as_evaled_expression(),
564:            return functools.partial(np.random.poisson, self["lam"].as_evaled_expression())
566:            p = self["p"].as_evaled_expression()
573:                self["n"].as_evaled_expression(),
574:                # self["p"].as_evaled_expression(),
579:                mean=self["mean"].as_evaled_expression(),
580:                sd=self["sd"].as_evaled_expression(),
581:                a=self["a"].as_evaled_expression(),
582:                b=self["b"].as_evaled_expression(),
586:                meanlog=self["meanlog"].as_evaled_expression(),
587:                sdlog=self["sdlog"].as_evaled_expression(),
595:            self.as_evaled_expression(),
596:            self.as_evaled_expression(),
```

emprzy committed:
Introduce a `_evaled_expression_int()` function

Implement this new function, add unit tests, rename other function to `_evaled_expression_float()`

emprzy committed:
Outsource distribution construction to `distributions.py`

introduce a new function (`build_distribution_from_confuse_config()`) that will assemble distributions accordingly

emprzy committed:
Combine `_pydantic_ext` expression evaluation functions, de-duplicate logic in `utils.py`

emprzy committed:
Fix tests

pearsonca commented on flepimop/gempyor_pkg/src/gempyor/utils copy.py:
seems to have been accidentally committed?

emprzy replied:
Definitely, thank you.

emprzy committed:
Remove accidental file duplicate

pearsonca commented on flepimop/gempyor_pkg/src/gempyor/utils.py:
```diff
@@ -551,49 +528,51 @@ def as_random_distribution(self):
         if dist == "fixed":
```
feels like all this stuff is duplicating the logic in distributions for converting from a confuse object? does it make sense to just replace the as_random_distribution calls and delete this method?

emprzy replied:
Hm definitely we need some de-duplication here. Let me figure out the optimal way to do it

emprzy committed:
Refactoring all uses of `.as_random_distribution` to `build_distribution_from_confuse_config()`

This code is presently broken.

emprzy committed:
Refactoring all uses of `.as_random_distribution` to `build_distribution_from_confuse_config()`

This code is presently broken.

emprzy committed:
Merge branch 'incorporate-gem.distributions-into-gem.parameters' of https://github.com/HopkinsIDD/flepiMoP into incorporate-gem.distributions-into-gem.parameters

pearsonca commented on flepimop/gempyor_pkg/src/gempyor/distributions.py:
```diff
@@ -51,6 +54,14 @@ def sample(
         rng = rng if rng is not None else self._rng
         return self._sample_from_generator(size=size, rng=rng)
+    
+    def __call__(self) -> float | int:
```
seems duplicated; aren't these going to be np.(float|int)64s?

emprzy replied:
which part seems duplicated?

TimothyWillard replied:
No, the `item` coerces a numpy array with a single value to `float | int`, https://numpy.org/doc/stable/reference/generated/numpy.ndarray.item.html.

pearsonca replied:
i'm literally seeing an exact copy of this __call__ declaration appear a few lines down; is that not actually there / some diff glitch?

emprzy committed:
De-duplicate code by removing three four functions from `gempyor.utils`

Add new `__call__` syntactic sugar to `Distributions`.

TimothyWillard commented on flepimop/gempyor_pkg/src/gempyor/_pydantic_ext.py:
```diff
@@ -8,16 +8,20 @@
 
 T = TypeVar("T")
 U = TypeVar("U")
+EE = TypeVar("EE", int, float)
```
I did not know about this `*constraints` usage, super helpful going forward!

emprzy replied:
Indeed, found it [here](https://docs.pydantic.dev/2.0/usage/types/typevars/)

TimothyWillard commented on flepimop/gempyor_pkg/src/gempyor/_pydantic_ext.py:
```diff
@@ -180,3 +184,64 @@ def _read_and_validate_dataframe(
+def _evaled_expression(val: Any, *, target_type: Type[float]) -> float | Any: ...
+
+
+def _evaled_expression(val: Any, *, target_type: Type[EE]) -> EE | Any:
```
```suggestion
@overload
def _evaled_expression(val: EE | str, *, target_type: Type[EE]) -> EE: ...

@overload
def _evaled_expression(val: T, *, target_type: Type[EE]) -> T: ...

def _evaled_expression(val: EE | str | Any, *, target_type: Type[EE]) -> EE | Any:
```
This should simplify the overloads and convey that when given `EE | str` the return is `EE` and otherwise it is as given.

TimothyWillard replied:
Also, is there a particular reason why the `target_type` is a keyword only argument? Seems like it just adds more typing for an internal function that has good documentation.

emprzy replied:
Okay sounds good, happy to get rid of `*`. I thought it would be a good clarity thing, to delineate between the target and what it returned (if they were not the same), but this would just become clear when an error gets thrown.

TimothyWillard commented on flepimop/gempyor_pkg/src/gempyor/_pydantic_ext.py:
```diff
@@ -8,16 +8,20 @@
 import pandas as pd
-from pydantic import BaseModel, RootModel
+from pydantic import BaseModel, RootModel, BeforeValidator
+import sympy
```
```suggestion
from sympy.parsing.sympy_parser import parse_expr
```
I think we only need this one thing from a specific module rather then the entire `sympy` package?

TimothyWillard commented on flepimop/gempyor_pkg/src/gempyor/_pydantic_ext.py:
```diff
@@ -180,3 +184,64 @@ def _read_and_validate_dataframe(
+        try:
+            return target_type(sympy.parsing.sympy_parser.parse_expr(val))
+        except TypeError as e:
+            raise ValueError(f"Can't convert expression to {target_type.__name__}.") from e
```
What value do we get out of wrapping the `TypeError` here, the exception doesn't catch syntax errors. i.e.
```
>>> sympy.parsing.sympy_parser.parse_expr("2..0 + 1")
ValueError: Error from parse_expr with transformed code: "Float ('2.' )Float ('.0' )+Integer (1 )"

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
...
    expr = eval(
           ^^^^^
  File "<string>", line 1
    Float ('2.' )Float ('.0' )+Integer (1 )
                 ^^^^^
SyntaxError: invalid syntax
```
Maybe:
```suggestion
        expr = parse_expr(val)
        if not expr.is_Number:
          raise ValueError(f"Cannot convert expression '{expr}' to {target_type}.")
        return target_type(expr)
```
Also open to other ideas. I'm not super familiar with `sympy` so this may not be the best way to do this.

TimothyWillard commented on flepimop/gempyor_pkg/src/gempyor/distributions.py:
```diff
@@ -52,6 +55,14 @@ def sample(
+
+    def __call__(self) -> float | int:
+        """A shortcut for `self.sample(size=1)`."""
+        return self.sample(size=1).item()
```
```suggestion
    def __call__(self) -> float | int:
        """A shortcut for `self.sample(size=1)`."""
        return self.sample(size=1).item()
```

TimothyWillard commented on flepimop/gempyor_pkg/src/gempyor/distributions.py:
```diff
@@ -476,3 +487,24 @@ def _sample_from_generator(
+DISTRIBUTION_ADAPTER = TypeAdapter(Distribution)
+
+
+def build_distribution_from_confuse_config(config: confuse.ConfigView) -> Distribution:
```
```suggestion
def distribution_from_confuse_config(config: confuse.ConfigView) -> Distribution:
```
I think this conveys the same meaning?

emprzy replied:
Nice, sounds good. I guess the word "build" implies something more involved is going on in the function than actually is.

TimothyWillard commented on flepimop/gempyor_pkg/tests/_pydantic_ext/test__evaled_expression.py:
```diff
@@ -0,0 +1,48 @@
+    ],
+)
+def test_evaled_expression_raises_value_error_for_invalid_input(val, target_type):
+    with pytest.raises(ValueError):
```
Should have a `match` to ensure we're getting the expected `ValueError`.

TimothyWillard commented on flepimop/gempyor_pkg/tests/_pydantic_ext/test__evaled_expression.py:
```diff
@@ -0,0 +1,48 @@
+from gempyor._pydantic_ext import _evaled_expression
+
+
+@pytest.mark.parametrize(
+    ("val", "target_type", "expected"),
+    [
+        ("1 + 1", float, 2.0),
+        ("5 / 2", float, 2.5),
+        ("sqrt(9)", float, 3.0),
+        (99.5, float, 99.5),
+        (-10.0, float, -10.0),
+        ("2 * 5", int, 10),
+        ("sqrt(81)", int, 9),
+        ("11 / 2", int, 5),
+        ("3.99", int, 3),
+        (99, int, 99),
+        (-50, int, -50),
+        (None, float, None),
+        (None, int, None),
+        (True, float, True),
+        (True, int, True),
+        (10.5, int, 10.5),
+        ([1, 2, 3], float, [1, 2, 3]),
+        ({"a": 1}, float, {"a": 1}),
+    ],
+)
```
Could this be parametrized by:
```
@pytest.mark.parametrize(("val", "expected"), ....)
@pytest.mark.parametrize("target_type"), [int, float])
```

emprzy replied:
how would this work...? Would this not create errors because the expected is the result of the `target_type`? The stacked pytest markers would (b/c stacking does cartesian product, i'm pretty sure) would test for `int` and `float` for each val/expected combination, which would cause some failures. Also, even if it would work, I'm not sure I understand the value that this would add.

TimothyWillard replied:
That's a good point, I was trying to figure out a way to deduplicate these tests:
```
        (None, float, None),
        (None, int, None),
        (True, float, True),
        (True, int, True),
        (10.5, int, 10.5),
        ([1, 2, 3], float, [1, 2, 3]),
        ({"a": 1}, float, {"a": 1}),
```
I guess you'd have to split this test into two, one when `val` is `str | int | float` and another test for everything else. So maybe not worth it in that case.

emprzy replied:
I don't understand what value doing so would add, besides potentially being less messy aesthetically and it being fractionally faster.

TimothyWillard replied:
The value would be deduplicating test cases. But given that it would require splitting the test two two it may not be wort it in this case.

TimothyWillard reviewed (changes requested):
Looks good overall, a handful of small edits.

pearsonca commented on flepimop/gempyor_pkg/tests/utils/test_utils2.py:
agree with removing things here, but: we need to *add* test somewhere (likely to the overarching `DistributionABC` generic tests) that those objects work as `Callable`s.

pearsonca reviewed (commented):
Mostly there, but we need to have a test for callability of DistributionABC

emprzy committed:
Update flepimop/gempyor_pkg/src/gempyor/_pydantic_ext.py

Co-authored-by: Timothy Willard <9395586+TimothyWillard@users.noreply.github.com>

emprzy committed:
Remove `*` keyword arg for `target_type` in `_evaled_expression()`

emprzy committed:
Update flepimop/gempyor_pkg/src/gempyor/_pydantic_ext.py

Co-authored-by: Timothy Willard <9395586+TimothyWillard@users.noreply.github.com>

emprzy committed:
Merge branch 'incorporate-gem.distributions-into-gem.parameters' of https://github.com/HopkinsIDD/flepiMoP into incorporate-gem.distributions-into-gem.parameters

emprzy committed:
Update flepimop/gempyor_pkg/src/gempyor/_pydantic_ext.py

Co-authored-by: Timothy Willard <9395586+TimothyWillard@users.noreply.github.com>

emprzy committed:
Merge branch 'incorporate-gem.distributions-into-gem.parameters' of https://github.com/HopkinsIDD/flepiMoP into incorporate-gem.distributions-into-gem.parameters

emprzy committed:
Naming suggestions, deleting duplication

emprzy committed:
Implement review suggestions

pearsonca commented on flepimop/gempyor_pkg/src/gempyor/NPI/SinglePeriodModifier.py:
```diff
@@ -152,12 +153,12 @@ def __createFromConfig(self, npi_config):
         if self.spatial_groups["grouped"]:
             for group in self.spatial_groups["grouped"]:
-                drawn_value = self.dist(size=1) * np.ones(len(group))
+                drawn_value = self.dist() * np.ones(len(group))
```
this is fine as-is, but seems clearer to implement as:

```suggestion
                drawn_value = np.repeat(self.dist(), len(group))
```

might also have some value in maintaining type more clearly?

TimothyWillard reviewed (approved):
Looks good to me

pearsonca commented on flepimop/gempyor_pkg/src/gempyor/parameters.py:
```diff
@@ -224,7 +230,9 @@ class via `subpop_names`.
             if "dist" in self.pdata[pn]:
-                param_arr[idx] = np.full((n_days, nsubpops), self.pdata[pn]["dist"]())
+                param_arr[idx] = np.full(
+                    (n_days, nsubpops), self.pdata[pn]["dist"].sample().item()
```
with the change to callability, can this now be

```suggestion
                    (n_days, nsubpops), self.pdata[pn]["dist"]()
```

pearsonca commented on flepimop/gempyor_pkg/src/gempyor/parameters.py:
```diff
@@ -272,8 +280,9 @@ def parameters_load(
-                pval = self.pdata[pn]["dist"]()
-                param_arr[idx] = np.full((n_days, nsubpops), pval)
+                param_arr[idx] = np.full(
+                    (n_days, nsubpops), self.pdata[pn]["dist"].sample().item()
```
same change due to callability?

emprzy committed:
Update flepimop/gempyor_pkg/src/gempyor/NPI/SinglePeriodModifier.py

Co-authored-by: Carl A. B. Pearson <pearsonca@users.noreply.github.com>

emprzy committed:
Update flepimop/gempyor_pkg/src/gempyor/parameters.py

Co-authored-by: Carl A. B. Pearson <pearsonca@users.noreply.github.com>

emprzy committed:
Update parameters.py
